### DISCUSSION TITLE
option to allow STDERR / open with command

### DISCUSSION THREAD

**Discussion by Docbroke** [leo-arch/clifm#69] [General]
STDERR of opener script seems to be redirected which is creating issues for me.
I am using opener script written in bash to open files. The script uses select loop, the output of which goes to STDERR by default ( I don't know why bash does that).

Here is an example of the problem.
```
#!/bin/bash

echo "hello"

select ops in 1 2 3 4 5; do
	echo $ops
done
```
Now let's run this redirecting STDOUT and STDERR alternatively
```
[ins] $ bash test 2> /dev/null
hello
^C

[ins] $ bash test 1> /dev/null
1) 1
2) 2
3) 3
4) 4
5) 5
#? ^C
```
output of echo goes to STDOUT but output of select loop goes to STDERR

Actual script I am using is [here](https://github.com/Docbroke/opener/blob/main/opener)
So when I use this as an opener in clifm, it's output asking to choose application is not displayed. As a workaround I added 2>&1 after all select loops, but still this doesn't work satisfactorily.

Can we allow display of STDERR as an option ?

**Comment by Docbroke** [comment]
main reason I am using this bash script is that it gives me "open with" choice. I guess that it may be desirable feature to add in lira, which will solve STDERR trouble for me ( excluding vis not opening with STDERR redirection ).
What I am suggesting here is Lira to ask user for choice from available applications to open given file, (open with)
```
$ o somefile.pdf
<autoopens file with default app>
$ ow somefile.pdf
OPEN somefile.pdf WITH:
     1. zathura
     2. mupdf-gl
     3. xournalpp
```

**Reply by leo-arch** [reply]
Hi @Docbroke. I'll take a look at the STDERR thing (shouldn't be any problem). As to the _Open With_ feature, it's a must. Don't know why I didn't think about this before. Thanks for suggesting it.

Btw, when I want to open a file with a specific application, I just append the executable name next the the ELN/filename:

```sh
somefile.pdf zathura
```

Of course, the other way round works as well:

```sh
zathura somefile.pdf
```

However, having an open-with command is a good idea: the command could suggest or list command names included in the corresponding line in the Lira config file (after checking commands existence).

**Comment by leo-arch** [comment]
I tried your script as opener (pretty nice btw), but cannot reproduce the issue: options are displayed either with STDERR silenced or not. What terminal are you using?

I'll add a command line switch so you can test it yourself.

Not sure if this is an issue with your script, but it refuses to run terminal applications, like `nano`. GUI stuff works fine.

EDIT: switch added. Run clifm with `--stderr`. This allows printing to STDERR when running the custom opener.

**Reply by Docbroke** [reply]
I had added `2>&1` after every select loop, therefor options are displayed with STDERR silenced. But still when I try to open clifm config file using function key shortcut it fails ( try yourself).
Issue with terminal program not working was already corrected but was not updated on github, I just did that now, you can check that. Nano shall work now.
I will check with --stderr and report.

**Reply by Docbroke** [reply]
I missed to answer your questions.
I am using sakura as my terminal.

**Comment by Docbroke** [comment]
latest version fails to build on alpine linux with this erros
```
src/file_operations.c: In function 'open_file':
src/file_operations.c:78:9: error: expected identifier before '(' token
   78 |   xargs.stderr ? E_NOFLAG : E_NOSTDERR) != EXIT_SUCCESS)
```
```
src/init.c: In function 'external_arguments':
src/init.c:1162:18: error: expected identifier before '(' token
 1162 |   case 45: xargs.stderr = 1; break;
      |                  ^~~~~~
src/init.c: In function 'unset_xargs':
src/init.c:1565:8: error: expected identifier before '(' token
 1565 |  xargs.stderr = UNSET;
```
```
make: *** [Makefile:33: build] Error 1
>>> ERROR: clifm: build failed
```

**Reply by leo-arch** [reply]
Not sure about that error. But it doesn't matter now: I temporarily removed the `--stderr` switch and added a `ow` command to allow the user to choose an opening application. It's not documented yet, but the syntax is just: `ow ELN/file`. You can add an ampersand at the end of the input string to run the program in the background.

This implementation is still not perfect (the `%f` placeholder is not honored and commands having arguments will lost them), but works.

Let me know what you think.

EDIT: You script works fine now.

**Reply by Docbroke** [reply]
Thanks, this is cool. I just compiled it. I think I won't need my script now, Lira has advantage of checking program availability. If I find issue I will report later.

**Reply by leo-arch** [reply]
Great! `ow` works now with commands with parameters and the `%f` placeholder.

**Comment by Docbroke** [comment]
```
<0> $ ow 23
1 sakura -f 28 -m -e epy.py
2 zathura
3 mupdf-gl
4 opener
5 xdg-open
Choose an application ('q' to quit): 1
sh: syntax error: unexpected "("
Thu, Oct 28, 16:14:49
```
same command works with `o 23` or `23` (default open)

Other command with parameters like `sakura -e vis` works with `ow`

**Reply by leo-arch** [reply]
What's the name of the file 23? Does it contains parenthesis? The problem is most likely that Lira is not quoting the file name. In that case, it's just a quick fix.

**Reply by Docbroke** [reply]
Yes indeed that file is having parenthesis in name, I just checked after seeing your message, other files opens fine. I was considering this as an issue due to spaces in command as same file opens fine with other options (like zathura)

**Reply by leo-arch** [reply]
This issue should be fixed now.

**Reply by Docbroke** [reply]
Yes, fixed.

**Comment by leo-arch** [comment]
The `ow` command supports now TAB completion. Check it out!
